chain/chain/src/chain.rs

```diff
@@ -1038,6 +1041,94 @@ impl Chain {
         }
     }
 
+    /// Check if any orphan with missing chunks is ready to request chunks
+    pub fn check_orphans_with_missing_chunks<F>(
+        &mut self,
+        me: &Option<AccountId>,
+        mut orphans_missing_chunks: F,
+    ) where
+        F: Copy + FnMut((CryptoHash, Vec<ShardChunkHeader>)) -> (),
+    {
+        let mut orphan_hashes = vec![];
+        for (orphan_hash, orphan) in self.orphans.orphans.iter() {
```

Comment:
Also probably limit this loop to some constant number of itarations (like 20), and prefer the entries from the end (`.iter().rev().take(n)`)
Better not to request chunks in advance than to constantly spend CPU cycles here if someone successfully spams the orphan pool.